controller/update.go

```diff
@@ -0,0 +1,255 @@
+	}
+	numAllowedToRemove := numRunning - opts.MinAlive
+
+	var numRemovalQueued int
+	for _, sliceID := range req.SliceIDs {
+		jobDesc := updateJobDesc{
+			Request: Request{
+				Group:    req.Group,
+				SliceIDs: []string{sliceID},
+				Units:    req.Units,
+			},
+		}
+
+		addQueue <- jobDesc
+
+		if numAllowedToRemove >= numRemovalQueued {
+			removeQueue <- jobDesc
+			numRemovalQueued++
+		}
+	}
+
+	return addQueue, removeQueue, nil
+}
+
+func (c controller) isGroupRemovalAllowed(req Request, opts UpdateOptions) (bool, error) {
+	numRunning, err := c.getNumRunningSlices(req)
+	if err != nil {
+		return false, maskAny(err)
+	}
+
+	if numRunning > opts.MinAlive {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (c controller) isGroupAdditionAllowed(req Request, numTotal int, opts UpdateOptions) (bool, error) {
+	numRunning, err := c.getNumRunningSlices(req)
+	if err != nil {
+		return false, maskAny(err)
+	}
+
+	if numRunning < numTotal+opts.MaxGrowth {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (c controller) addWorker(jobDesc updateJobDesc, removeQueue chan updateJobDesc, fail chan<- error) {
+	// Submit.
+	taskObject, err := c.Submit(jobDesc.Request)
+	if err != nil {
+		fail <- maskAny(err)
+		return
+	}
+	closer := make(<-chan struct{})
+	taskObject, err = c.WaitForTask(taskObject.ID, closer)
+	if err != nil {
+		fail <- maskAny(err)
+		return
+	}
+	if task.HasFailedStatus(taskObject) {
+		fail <- maskAny(fmt.Errorf(taskObject.Error))
+		return
+	}
+
+	// Start.
+	taskObject, err = c.Start(jobDesc.Request)
+	if err != nil {
+		fail <- maskAny(err)
+		return
+	}
+	closer = make(<-chan struct{})
+	taskObject, err = c.WaitForTask(taskObject.ID, closer)
+	if err != nil {
+		fail <- maskAny(err)
+		return
+	}
+	if task.HasFailedStatus(taskObject) {
+		fail <- maskAny(fmt.Errorf(taskObject.Error))
+		return
+	}
+
+	removeQueue <- jobDesc
+}
+
+func (c controller) removeWorker(jobDesc updateJobDesc, fail chan<- error) {
+	// Stop.
+	taskObject, err := c.Stop(jobDesc.Request)
+	if err != nil {
+		fail <- maskAny(err)
+		return
+	}
+	closer := make(<-chan struct{})
+	taskObject, err = c.WaitForTask(taskObject.ID, closer)
+	if err != nil {
+		fail <- maskAny(err)
+		return
+	}
+	if task.HasFailedStatus(taskObject) {
+		fail <- maskAny(fmt.Errorf(taskObject.Error))
+		return
+	}
+
+	// Destroy.
+	taskObject, err = c.Destroy(jobDesc.Request)
+	if err != nil {
+		fail <- maskAny(err)
+		return
+	}
+	closer = make(<-chan struct{})
+	taskObject, err = c.WaitForTask(taskObject.ID, closer)
+	if err != nil {
+		fail <- maskAny(err)
+		return
+	}
+	if task.HasFailedStatus(taskObject) {
+		fail <- maskAny(fmt.Errorf(taskObject.Error))
+		return
+	}
+}
+
+// UpdateOptions represents the options defining the strategy of an update
+// process. Lets have a look at how the update process of 3 group slices would
+// look like using the given options.
+//
+//     TODO I am not that happy with this visualization. Improving it? Removing it?
```

Comment:
Yeah getting the time stream right is not really possible here. At least I don't know how.